unstructured/ingest/connector/chroma.py

```diff
@@ -0,0 +1,162 @@
+                settings=self.connector_config.settings,
+                tenant=self.connector_config.tenant,
+                database=self.connector_config.database,
+            )
+        else:
+            raise ValueError("Chroma connector requires either path or host and port to be set.")
+
+        collection = chroma_client.get_or_create_collection(
+            name=self.connector_config.collection_name
+        )
+        return collection
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["chromadb"], extras="chroma")
+    def upsert_batch(self, batch):
+        collection = self.chroma_collection
+
+        try:
+            # Chroma wants lists even if there is only one element
+            collection.add(
+                ids=batch["ids"],
+                documents=batch["documents"],
+                embeddings=batch["embeddings"],
+                metadatas=batch["metadatas"],
+            )
+        except Exception as e:
+            raise ValueError(f"chroma error: {e}") from e
+
+    @staticmethod
+    def chunks(iterable, batch_size=100):
+        """A helper function to break an iterable into chunks of size batch_size."""
+        it = iter(iterable)
+        chunk = tuple(itertools.islice(it, batch_size))
+        while chunk:
+            yield chunk
+            chunk = tuple(itertools.islice(it, batch_size))
+
+    @staticmethod
+    def prepare_chroma_dict(chunk: t.Tuple[t.Dict[str, t.Any]]) -> t.Dict[str, t.List[t.Any]]:
+        """Helper function to break a tuple of dicts into list of parallel lists for ChromaDb.
+        ({'id':1}, {'id':2}, {'id':3}) -> {'ids':[1,2,3]}"""
+        chroma_dict = {}
+        chroma_dict["ids"] = [x.get("id") for x in chunk]
+        chroma_dict["documents"] = [x.get("document") for x in chunk]
+        chroma_dict["embeddings"] = [x.get("embedding") for x in chunk]
+        chroma_dict["metadatas"] = [x.get("metadata") for x in chunk]
+        # Make sure all lists are of the same length
+        assert (
+            len(chroma_dict["ids"])
+            == len(chroma_dict["documents"])
+            == len(chroma_dict["embeddings"])
+            == len(chroma_dict["metadatas"])
+        )
+        return chroma_dict
+
+    def write_dict(self, *args, dict_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(f"Inserting / updating {len(dict_list)} documents to destination ")
+
+        chroma_batch_size = self.write_config.batch_size
+
+        for chunk in self.chunks(dict_list, chroma_batch_size):
+            self.upsert_batch(self.prepare_chroma_dict(chunk))
+
+    def write(self, docs: t.List[BaseIngestDoc]) -> None:
+        dict_list: t.List[t.Dict[str, t.Any]] = []
+        for doc in docs:
+            local_path = doc._output_filename
+            with open(local_path) as json_file:
+                dict_content = json.load(json_file)
+
+                dict_content = [
+                    {
```

Comment:
Maybe we could unite this dict parsing part with the function `prepare_chroma_dict` above, putting all dict forming logic at one place

Reply:
I like the thought. Unfortunately then the chunks method and multiprocessing (which will work eventually) become much more complicated because you have a bunch of parallel lists instead of a list of dicts. (i changed prepare_chroma_dict to prepare_chroma_lists) So I'd like to keep them separate.

Reply:
Sorry, rethinking, what you do does make sense. Agreed with all